src/main/java/org/opensearch/geospatial/index/mapper/shape/ShapeIndexer.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.index.mapper.shape;
+
+import static org.opensearch.geometry.ShapeType.CIRCLE;
+import static org.opensearch.geometry.ShapeType.LINEARRING;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.lucene.document.XYShape;
+import org.apache.lucene.geo.XYLine;
+import org.apache.lucene.geo.XYPolygon;
+import org.apache.lucene.index.IndexableField;
+import org.opensearch.geometry.Circle;
+import org.opensearch.geometry.Geometry;
+import org.opensearch.geometry.GeometryCollection;
+import org.opensearch.geometry.GeometryVisitor;
+import org.opensearch.geometry.Line;
+import org.opensearch.geometry.LinearRing;
+import org.opensearch.geometry.MultiLine;
+import org.opensearch.geometry.MultiPoint;
+import org.opensearch.geometry.MultiPolygon;
+import org.opensearch.geometry.Point;
+import org.opensearch.geometry.Polygon;
+import org.opensearch.geometry.Rectangle;
+import org.opensearch.geometry.ShapeType;
+import org.opensearch.geospatial.index.common.shape.ShapeConverter;
+import org.opensearch.index.mapper.AbstractGeometryFieldMapper;
+import org.opensearch.index.mapper.ParseContext;
+
+/**
+ * Converts geometries into Lucene-compatible form for indexing in a shape field.
+ */
+public class ShapeIndexer implements AbstractGeometryFieldMapper.Indexer<Geometry, Geometry> {
+
+    private final String name;
+    private final GeometryVisitor<IndexableField[], RuntimeException> visitor;
+
+    public ShapeIndexer(String name) {
+        this.name = Objects.requireNonNull(name, "name cannot be null");
+        visitor = new XYShapeGeometryVisitor(name);
+    }
+
+    @Override
+    public Geometry prepareForIndexing(Geometry geometry) {
+
+        if (geometry == null) {
+            return null;
+        }
+        return geometry.visit(new GeometryVisitor<>() {
+            public Geometry visit(Circle circle) {
```

Comment:
why we are not creating another class for this implementation of GeometryVisitor?

Reply:
Just to keep consistent with GeoShape. May be we don't have to. I moved to new file now.